runtime/lua/vim/lsp/handlers.lua

```diff
@@ -197,7 +197,11 @@ M['textDocument/references'] = function(_, result, ctx, config)
         items = util.locations_to_items(result, client.offset_encoding),
         context = ctx,
       })
-      api.nvim_command('botright copen')
+      if not config.no_qf_window then
+        api.nvim_command('botright copen')
+      else
+        api.nvim_command('cfirst')
+      end
```

Comment:
if we're going to introduce new surface area, it should be a function, not another hardcoded thing. Something like this:

```suggestion
      if not config.on_after_setqflist then
        api.nvim_command('botright copen')
      else
        assert(type(config.on_after_setqflist) == 'function', 'on_after_setqflist is not a function')
        config.on_after_setqflist()
      end
```

And this should be done for the loclist case as well. But the option name is just `on_after_setqflist`.

We should have generalized `config.loclist` to `config.on_qfitems`, but too late now.

Reply:
Maybe it is not too late and we can introduce breaking change with deprecation warning?

Reply:
`config.on_qfitems` would be a more complex discussion. I think `config.on_qflist` is a reasonable change given the number of places where it is relevant, but also want to hear from @neovim/lsp

Reply:
Something like `on_qflist` sounds good to me I think we shouldn't extend `config` with it but instead add `options` to `buf.references`. For the reasons outlined here https://github.com/neovim/neovim/pull/18577#discussion_r873420834

Reply:
> but I think we shouldn't extend `config` with it and instead add `options` to `buf.references`. See the reasons outlined here [#18577 (comment)](https://github.com/neovim/neovim/pull/18577#discussion_r873420834)

Reply:
@mfussenegger does this PR look like what you expect?